src/ORM/DataObject.php

```diff
@@ -316,88 +335,107 @@ public static function getSchema()
+                    if (!array_key_exists($field, $record)) {
+                        $this->record[$field . '_Lazy'] = $fieldClass;
+                    }
+                }
 
-            user_error(
-                "DataObject::__construct passed $passed.  It's supposed to be passed an array,"
-                . " taken straight from the database.  Perhaps you should use DataList::create()->First(); instead?",
-                E_USER_WARNING
-            );
-            $record = null;
-        }
+                $this->original = $this->record;
+                $this->changed = [];
+                $this->changeForced = false;
 
-        // Set $this->record to $record, but ignore NULLs
-        $this->record = array();
-        foreach ($record as $k => $v) {
-            // Ensure that ID is stored as a number and not a string
-            // To do: this kind of clean-up should be done on all numeric fields, in some relatively
-            // performant manner
-            if ($v !== null) {
-                if ($k == 'ID' && is_numeric($v)) {
-                    $this->record[$k] = (int)$v;
-                } else {
-                    $this->record[$k] = $v;
+                break;
+
+            // Create a new object, using the constructor argument as the initial content
+            case self::CREATE_OBJECT:
+                if ($record instanceof stdClass) {
+                    $record = (array)$record;
                 }
-            }
-        }
 
-        // Identify fields that should be lazy loaded, but only on existing records
-        if (!empty($record['ID'])) {
-            // Get all field specs scoped to class for later lazy loading
-            $fields = static::getSchema()->fieldSpecs(
-                static::class,
-                DataObjectSchema::INCLUDE_CLASS | DataObjectSchema::DB_ONLY
-            );
-            foreach ($fields as $field => $fieldSpec) {
-                $fieldClass = strtok($fieldSpec, ".");
-                if (!array_key_exists($field, $record)) {
-                    $this->record[$field . '_Lazy'] = $fieldClass;
+                if (!is_array($record)) {
+                    if (is_object($record)) {
+                        $passed = "an object of type '" . get_class($record) . "'";
+                    } else {
+                        $passed = "The value '$record'";
+                    }
+
+                    user_error(
+                        "DataObject::__construct passed $passed.  It's supposed to be passed an array,"
+                        . " taken straight from the database.  Perhaps you should use DataList::create()->First(); instead?",
+                        E_USER_WARNING
+                    );
+                    $record = [];
                 }
-            }
-        }
 
-        $this->original = $this->record;
+                // Default columns
+                $this->record['ID'] = empty($record['ID']) ? 0 : $record['ID'];
```

Comment:
This used to cast to `(int)` - did you remove that on purpose?

Reply:
I think my rationale was No longer needed thanks to my strict-typing work